src/include/sof/drivers/sai.h

```diff
@@ -0,0 +1,233 @@
+#define REG_SAI_TFR4    0x50 /* SAI Transmit FIFO */
+#define REG_SAI_TFR5    0x54 /* SAI Transmit FIFO */
+#define REG_SAI_TFR6    0x58 /* SAI Transmit FIFO */
+#define REG_SAI_TFR7    0x5C /* SAI Transmit FIFO */
+#define REG_SAI_TMR	0x60 /* SAI Transmit Mask */
+#define REG_SAI_TTCTL	0x70 /* SAI Transmit Timestamp Control Register */
+#define REG_SAI_TTCTN	0x74 /* SAI Transmit Timestamp Counter Register */
+#define REG_SAI_TBCTN	0x78 /* SAI Transmit Bit Counter Register */
+#define REG_SAI_TTCAP	0x7C /* SAI Transmit Timestamp Capture */
+
+#define REG_SAI_RCSR	0x80 /* SAI Receive Control */
+#define REG_SAI_RCR1	0x84 /* SAI Receive Configuration 1 */
+#define REG_SAI_RCR2	0x88 /* SAI Receive Configuration 2 */
+#define REG_SAI_RCR3	0x8c /* SAI Receive Configuration 3 */
+#define REG_SAI_RCR4	0x90 /* SAI Receive Configuration 4 */
+#define REG_SAI_RCR5	0x94 /* SAI Receive Configuration 5 */
+#define REG_SAI_RDR0    0xa0 /* SAI Receive Data */
+#define REG_SAI_RDR1    0xa4 /* SAI Receive Data */
+#define REG_SAI_RDR2    0xa8 /* SAI Receive Data */
+#define REG_SAI_RDR3    0xac /* SAI Receive Data */
+#define REG_SAI_RDR4    0xb0 /* SAI Receive Data */
+#define REG_SAI_RDR5    0xb4 /* SAI Receive Data */
+#define REG_SAI_RDR6    0xb8 /* SAI Receive Data */
+#define REG_SAI_RDR7    0xbc /* SAI Receive Data */
+#define REG_SAI_RFR0    0xc0 /* SAI Receive FIFO */
+#define REG_SAI_RFR1    0xc4 /* SAI Receive FIFO */
+#define REG_SAI_RFR2    0xc8 /* SAI Receive FIFO */
+#define REG_SAI_RFR3    0xcc /* SAI Receive FIFO */
+#define REG_SAI_RFR4    0xd0 /* SAI Receive FIFO */
+#define REG_SAI_RFR5    0xd4 /* SAI Receive FIFO */
+#define REG_SAI_RFR6    0xd8 /* SAI Receive FIFO */
+#define REG_SAI_RFR7    0xdc /* SAI Receive FIFO */
+#define REG_SAI_RMR	0xe0 /* SAI Receive Mask */
+#define REG_SAI_RTCTL	0xf0 /* SAI Receive Timestamp Control Register */
+#define REG_SAI_RTCTN	0xf4 /* SAI Receive Timestamp Counter Register */
+#define REG_SAI_RBCTN	0xf8 /* SAI Receive Bit Counter Register */
+#define REG_SAI_RTCAP	0xfc /* SAI Receive Timestamp Capture */
+
+#define REG_SAI_MCTL	0x100 /* SAI MCLK Control Register */
+#define REG_SAI_MDIV	0x104 /* SAI MCLK Divide Register */
+
+#define REG_SAI_XCSR(tx)	(tx ? REG_SAI_TCSR : REG_SAI_RCSR)
+#define REG_SAI_XCR1(tx)	(tx ? REG_SAI_TCR1 : REG_SAI_RCR1)
+#define REG_SAI_XCR2(tx)	(tx ? REG_SAI_TCR2 : REG_SAI_RCR2)
+#define REG_SAI_XCR3(tx)	(tx ? REG_SAI_TCR3 : REG_SAI_RCR3)
+#define REG_SAI_XCR4(tx)	(tx ? REG_SAI_TCR4 : REG_SAI_RCR4)
+#define REG_SAI_XCR5(tx)	(tx ? REG_SAI_TCR5 : REG_SAI_RCR5)
+#define REG_SAI_XMR(tx)		(tx ? REG_SAI_TMR : REG_SAI_RMR)
+
+#define REG_SAI_XMR_MASK        0xffffffff
+
+/* SAI Transmit/Receive Control Register */
+#define REG_SAI_CSR_TERE	BIT(31)
+#define REG_SAI_CSR_SE		BIT(30)
+#define REG_SAI_CSR_FR		BIT(25)
+#define REG_SAI_CSR_SR		BIT(24)
+#define REG_SAI_CSR_XF_SHIFT	16
+#define REG_SAI_CSR_XF_W_SHIFT	18
+#define REG_SAI_CSR_XF_MASK	(0x1f << REG_SAI_CSR_XF_SHIFT)
```

Comment:
You can use `MASK` macro.

Reply:
I find MASK macro less straightforward, i'd rather stick with shifts

Reply:
@groncarolo This one would be rewritten as MASK(22, 18) (the bits 22 through 18, inclusive). See an extended list of examples in my EDMA driver (as well as my not-yet-cleaned-up ESAI driver).

It took me a few days to get used to this macro (and the SET_BITS one, which is related) but when I got them intuitively I get why they are in fact preferred.

Reply:
done